### PR TITLE
More Advanced API Naming, Tests, and Stability

### PR DESCRIPTION
**Pull request by nleroy917**
## Overview
These changes are all-encompassing to the package and I was just trying to mature the API naming convention, build up some tests, and increase stability to handle edge cases, errors, etc. I also tried to create a consistent way of making database queries.

Most of my changes were made with pephub in mind. So, I tried to make returned results practical for something like that. If we need to be less specific, we can.

## Changes
I made 3 main changes to the codebase:

1. **Standardize the API naming discussed in #18**
We now have 4 main getters: `get_project`, `get_projects`, `get_namespace`, and `get_namespaces`.  For the **plural** getters, they can accept **no input** or **list input**. For **no input**, the function will retrieve all available objects. For the **list input**, it will just return those objects named in the list. The **singular** functions will return just one instance of a project specified by a registry path (`get_project`) or `str` (`get_namespace`).

2. **I updated the internal `sql` query function** 
For [parameterized queries](https://www.psycopg.org/psycopg3/docs/basic/params.html) it was accepting an unspecified number of unnamed arguments, so I just changed it to accept a `tuple` as `psycopg` wants. There is some internal logic to map single values or a `list` to a `tuple` if not supplied properly.

3. **Unit tests**
With `pytest` I added some pretty basic unit test functionality. Was very helpful for testing my code and debugging. With this, I removed the big `main` block that was at the bottom of the `pepagent.py` file.

I also did some small things like rename an exception, moved constants around, and created some utils in `utils.py`.

## Dependencies
I didn't add any dependencies, but we are heavily reliant on the development version of `peppy`, so before this can really get released, we probably need a release of `peppy`.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "python.testing.pytestArgs": [
+        "tests"
+    ],
+    "python.testing.unittestEnabled": false,
+    "python.testing.pytestEnabled": true
+}
```
Comment (Khoroshevskyi) [inline .vscode/settings.json:7]: add .vscode to gitignore, and delete this folder

File: pepagent/pepagent.py [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
 import peppy
 import os
 from hashlib import md5
+from itertools import chain
 import ubiquerg
+from yaml import parse
+
+from pepagent.utils import is_valid_resgistry_path
+
 from .exceptions import *
```
Comment (Khoroshevskyi) [inline pepagent/pepagent.py:16]: Delete this line

File: pepagent/pepagent.py [hunk 3]
```diff
@@ -304,31 +402,45 @@ def _get_namespace_proj_anno(self, namespace: str = None) -> dict:
...
+                return list(output_result)
+            else:
+                return None
         except psycopg2.Error as e:
             _LOGGER.error(f"Error occurred while running query: {e}")
         finally:
             cursor.close()
 
-    def run_sql_search_all(self, sql_query: str, *argv) -> list:
+    def run_sql_fetchall(self, sql_query: str, argv=()) -> list:
```
Comment (Khoroshevskyi) [inline pepagent/pepagent.py:432]: Why did you change *args to args as set?

File: pepagent/pepagent.py [hunk 4]
```diff
@@ -304,31 +402,45 @@ def _get_namespace_proj_anno(self, namespace: str = None) -> dict:
...
         """
         Fetching all result by providing sql query and arguments
         :param str sql_query: sql string that has to run
         :param argv: arguments that has to be added to sql query
```
Comment (Khoroshevskyi) [inline pepagent/pepagent.py:436]: If you changes *args , change documentation here too

**Review by Khoroshevskyi — COMMENTED** [review]
I like it! But also have few comments.

File: pepagent/pepagent.py (hunk 3 above)
Comment (nleroy917) [reply pepagent/pepagent.py:432]: Yeah, so `psycopg` wants a tuple, so I thought we should be passing a tuple as well to the `run_sql_fetchall` function... I am not sure if this **truly** is best. A variable-length list of unnamed args after the `sql_query` arg seemed dicey to me. More granular control over a tuple. I guess, why not just be explicit in the way we pass in the parameters for the query.

I am open to reverting, though. Maybe @nsheff has a preference.

Comment (nleroy917) [reply pepagent/pepagent.py:432]: Ok. Reading more into this, I think your approach was better. After looking through this thread on SO I am more partial to `*args`: https://stackoverflow.com/questions/33542959/why-use-packed-args-kwargs-instead-of-passing-list-dict

I'll revert back today and make sure all tests still pass.

Comment (nleroy917) [reply pepagent/pepagent.py]: Fixed.

File: pepagent/exceptions.py [hunk 5]
```diff
@@ -1,10 +1,10 @@
 """ Custom error types """
 
 
-__all__ = ["PEPDBCorrectnessError"]
+__all__ = ["SchemaError"]
```
[the remark this answers is not in the document]
Comment (nleroy917) [reply pepagent/exceptions.py:4]: I think this was Alex. @Khoroshevskyi

File: pepagent/pepagent.py [hunk 6]
```diff
@@ -171,45 +166,145 @@ def get_project(
...
+        projects = list(
+            map(
+                lambda p: {
+                    "id": p[0],
+                    "name": p[1],
+                    "digest": p[2],
+                    "description": p[3]["proj_description"],
+                    "n_samples": p[3]["n_samples"],
+                },
+                results,
```
[the remark this answers is not in the document]
Comment (rafalstepien) [reply pepagent/pepagent.py:271]: Exactly, and even if you will put the data in `results` list to for example `pydantic` model, you could do something like:

```
[
  {
      "id": result.id,
      "name": result.name,
      "digest": result.digest,
      "description": result.description.proj_description,
      "n_samples": result.n_samples.n_samples,
  } for result in results
]
```

Which is much cleaner than referencing by index. But this is obviously a little bit more work and maybe not for this pull request. Anyway worth considering.

**Comment by nleroy917** [comment]
Tests are failing because there is no database to test against. See here: #21
